File: lc_soln/0002_add_two_numbers/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution
{
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
    {
        var dummy = new ListNode();
        ListNode curr = dummy;
        int total = 0;
        while (l1 != null || l2 != null || total > 0)
        {
            if (l1 != null)
            {
                total += l1.val;
                l1 = l1.next;
            }
            if (l2 != null)
            {
                total += l2.val;
                l2 = l2.next;
            }
            var node = new ListNode(total % 10);
            total /= 10;
            curr.next = node;
            curr = curr.next;
        }
        return dummy.next;
    }
}

File: lc_soln/0002_add_two_numbers/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummy = ListNode()
        curr = dummy
        total = 0
        while l1 is not None and l2 is not None:
            total += l1.val + l2.val
            l1 = l1.next
            l2 = l2.next
            node = ListNode(val=total % 10)
            total //= 10
            curr.next = node
            curr = curr.next
        while l1 is not None:
            total += l1.val
            l1 = l1.next
            node = ListNode(val=total % 10)
            total //= 10
            curr.next = node
            curr = curr.next
        while l2 is not None:
            total += l2.val
            l2 = l2.next
            node = ListNode(val=total % 10)
            total //= 10
            curr.next = node
            curr = curr.next
        if total > 0:
            curr.next = ListNode(val=total)
        return dummy.next


def main(): ...


if __name__ == "__main__":
    main()
